store/etcd/rulesets.go

```diff
@@ -55,7 +63,46 @@ func (s *RulesetService) List(ctx context.Context, prefix string, limit int, con
 
 	options = append(options, clientv3.WithLimit(int64(limit)))
 
-	resp, err := s.Client.KV.Get(ctx, s.rulesetsPath(key, ""), options...)
+	if pathsOnly {
+		return s.listPaths(ctx, key, prefix, limit, options)
+	}
+	return s.listRulesets(ctx, key, prefix, limit, options)
+}
+
+func (s *RulesetService) listPaths(ctx context.Context, key, prefix string, limit int, opts []clientv3.OpOption) (*store.RulesetEntries, error) {
```

Comment:
This function is the same as `listRulesets`. The only differences are:
- the options
- the complete key passed to `s.Client.KV.Get`
- what is done of a pair

How would you factorize that?

Reply:
... and the generated `continue` token.

honestly, I don't see how can I do better?

Reply:
You could have something like that:

```go
// line 71

if pathsOnly {
  opts = append(opts, clientv3.WithKeysOnly())
  entries, err = s.list(ctx, s.latestRulesetPath(key), opts, func(pair mvccpb.KeyValue) *store.RulesetEntry {
    p := strings.TrimPrefix(string(pair.Key), s.latestRulesetPath("")+"/")
    return &store.RulesetEntry{Path: p}, nil
  })
  if err == nil && len(entries) > 0 {
    entries.Continue = base64.URLEncoding.EncodeToString([]byte(entries[len(entries)-1].Path + "\x00"))
  }
} else {
  ...
}

if err != nil {
  return nil, blabla
}
```

Reply:
Nevermind, it's not worth it because of the `continue` token. Removing it would remove a lot of code from the `listRulesets` method which makes it not interesting anymore to factorize

Reply:
Exactly. And we will end with a lot of `if pathsOnly { ... } else { ... }` blocks.